aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -1337,6 +1337,104 @@ void qadaptive_avg_pool2d_nhwc_kernel(
   });
 }
 
+void qadaptive_avg_pool3d_ndhwc_kernel(
+    const Tensor& qx,
+    Tensor& qy,
+    int64_t b,
+    int64_t sizeC,
+    int64_t isizeD,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeD,
+    int64_t osizeH,
+    int64_t osizeW,
+    int64_t istrideB,
+    int64_t istrideC,
+    int64_t istrideD,
+    int64_t istrideH,
+    int64_t istrideW) {
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "adaptive_avg_pool3d_ndhwc", [&]() {
+    scalar_t* idata = static_cast<scalar_t*>(qx.data_ptr());
+    scalar_t* odata = static_cast<scalar_t*>(qy.data_ptr());
+    auto minimum = std::numeric_limits<scalar_t::underlying>::lowest();
+    auto maximum = std::numeric_limits<scalar_t::underlying>::max();
+    auto* i_p =
+        reinterpret_cast<typename scalar_t::underlying*>(idata + b * istrideB);
+
+    float input_scale = qx.q_scale();
+    float output_scale = qy.q_scale();
+    int input_zero_point = qx.q_zero_point();
+    int output_zero_point = qy.q_zero_point();
+
+    for (int64_t od = 0; od < osizeD; od++) {
+      int istartD = (int)std::floor((float)(od * isizeD) / osizeD);
+      int iendD = (int)std::ceil((float)((od + 1) * isizeD) / osizeD);
+      int kD = iendD - istartD;
+      for (int64_t oh = 0; oh < osizeH; oh++) {
+        int istartH = (int)std::floor((float)(oh * isizeH) / osizeH);
+        int iendH = (int)std::ceil((float)((oh + 1) * isizeH) / osizeH);
+        int kH = iendH - istartH;
+        for (int64_t ow = 0; ow < osizeW; ow++) {
+          auto* o_p = reinterpret_cast<typename scalar_t::underlying*>(
+              odata + b * osizeD * osizeH * osizeW * sizeC +
+              (od * osizeH * osizeW + oh * osizeW + ow) * sizeC);
+          int istartW = (int)std::floor((float)(ow * isizeW) / osizeW);
+          int iendW = (int)std::ceil((float)((ow + 1) * isizeW) / osizeW);
+          int kW = iendW - istartW;
+          int size = kD * kH * kW;
+          float multiplier = input_scale / output_scale / size;
+          int input_zero_point_m_size = -input_zero_point * size;
+          int64_t c = 0;
+          // For int8 or uint8quantization, we implicitly use int32 as
+          // accumulation Or else, it will go to the slow path
+          // TODO: support 16bit, 32bit, and etc.
+          auto* internal_i_p = i_p + istartH * istrideH + istartW * istrideW;
+
+          // TODO: more vectorization with loop interleaving
+          do_avg_pool_on_AVX2<scalar_t>(
```

Comment:
not related to this diff, but it's non intuitive that this function is a no-op if AVX2 is not available.  Maybe we can add a brief comment here to explain?